enhancements/baremetal/baremetal-poison-pill-remediation.md

```diff
@@ -0,0 +1,189 @@
+1. Mark the node as unschedulable (this prevents the node to run new workload after reboot)
+1. Add current time to annotation
+1. Unhealthy node reboot itself
+1. After *timeout-to-assume-node-rebooted* (either configurable or calculated by software) - delete the unhealthy node (to signal the cluster that the workload is no longer running there, and it is safe to run it elsewhere)
+1. Recreate the node object, without the unschedulable taint
+
+The first step of marking the node as unschedulable is important, as the reboot might fix the health issue, and the scheduler could assign workload to that node after the it rebooted, which could result in other nodes deleteing the rebooted node while it is running some workload. To avoid that, the others will delete it and report it as unhealthy only after marking it as unschedulable which ensures it will not get any workloads after reboot.
+
+### User Stories [optional]
+
+#### Story 1
+As a non-BMC baremetal cluster admin, I would like to have automated self-healing nodes. This will make my cluster more robust and immune to failures
+
+#### Story 2
+As a non-BMC baremetal cluster admin, I would like to minimze downtime for my applications, even if there's a node failure.
+
+### Implementation Details/Notes/Constraints [optional]
+
+This enhancement relies on the existence of MHC, which is currently working only in IPI and assisted insaller.
+We might want to enable MHC for UPI clusters as well, maybe by faking Machine CRs for the nodes.
+
+### Risks and Mitigations
+
+If the unhealthy node doesn't reboot in time, then there's a risk that a node will be deleted while the workload is still running.
+To mitigate this we prefer a hardware watchdog which ensures that reboot will take place if the poison pill is not running.
+
+There could be cases where this algorithm will declare a node as unhealthy while it’s healthy (false negative)
+For example, if someone made some network configuration mistake which blocks communication between the nodes - all nodes will consider themselves unhealthy and remediation will be triggered.
+In this case, the cluster is mostly in an unusable state anyway, so no significant harm is made.
+
+
+## Design Details
+
+### Upgrade / Downgrade Strategy
+
+N/A. This is a new component.
+
+### Version Skew Strategy
+
+Downgrade from a version with poison pill to a version without it during remediation process can result in a node that doesn't exist in api-server.
+As part of the remediation process, the poison pill agents keep a backup for the node yaml.
+They are expected to use that backup after the node has been deleted, to re-create the node.
+If a downgrade happens during that process, there will be no poison pill agents to re-create the node.
+
+In addition, a node might consult with its peers for its health status. If they are downgraded to a version without poison piil,
+they won't respond, and we might get false positive.
+
+## Drawbacks
+
+TBD
```

Comment:
We mention a few above with False-Negatives, should we explore / outline more of those here?